### DISCUSSION TITLE
Why not use luagit2?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [SuperBo/fugit2.nvim#96] [Q&A] [answered]
libgit2 has a official lua binding: https://github.com/libgit2/luagit2

Why this project don't use it?

**Comment by Freed-Wu** [comment]
Looks like luagit2 only support lua 5.2+ until now https://github.com/libgit2/luagit2/issues/7

**Comment by SuperBo** [comment, marked as answer]
Hi @Freed-Wu, as you can see in repo's history, it's is just recently updated. The previous update is from 4 years ago.

![Screenshot 2024-09-24 at 13 19 13](https://github.com/user-attachments/assets/3d24841d-7413-4ab9-b76c-4125e8c6f98b)

Neovim use luajit, so I want to make use of LuaJit FFI  https://luajit.org/ext_ffi.html to make most speed of lua.

**Reply by Freed-Wu** [reply]
Oh, I tend to use official lua binding to save the effort to adapt upstream's API. However don't know why recently they restart development (May be a good message for downstream? If official binding is enough stable, may this project try it?)

About performance, I am not sure ffi v.s. lua bindings in C. IMO, FFI save the effort to write C code, however, in performance, C should be better?

**Reply by SuperBo** [reply]
@Freed-Wu, quote from introduction of LuaJIT FFI.

`The FFI library is tightly integrated into LuaJIT (it's not available as a separate module). The code generated by the JIT-compiler for accesses to C data structures from Lua code is on par with the code a C compiler would generate. Calls to C functions can be inlined in JIT-compiled code, unlike calls to functions bound via the classic Lua/C API. `

Another advantage of FFI is garbage collector support, so we can assure memory safety when we forgot to free unused memory.

**Reply by Freed-Wu** [reply]
https://neovim.io/doc/user/lua.html

looks like lua 5.1 is always available, luajit is not guaranteed?

**Reply by SuperBo** [reply]
@Freed-Wu another advantage I can think of is user don't need to run compile step when using FFI. With normal lua binding, a compile step must be done to compile C binding header.

And I designed this lib with LuaJIT in mind first :(

**Reply by Freed-Wu** [reply]
> user don't need to run compile step

Luarocks allow distribute compiled binary rocks (lua bindings for libgit2) so it shouldn't be a problem except they use unusual OSes :smile:

**Reply by Freed-Wu** [reply]
Another case is in some OSes, the libgit2's path is not fixed. Such as NixOS's `/nix/store/qra2k11wd0pzwbsqraml3y57v6gy37ys-libgit2-1.8.1-lib/lib/libgit2.so`. If we let user config it, it will be a little troublesome. However, build it as a lua module can write the path to dynamic link library's `rpath` to save user's config. such as:

```sh
$ ldd ~/.local/lib/lua/5.1/tellenc.so
	linux-vdso.so.1 (0x00007fea683bc000)
	libstdc++.so.6 => /nix/store/22nxhmsfcv2q2rpkmfvzwg2w5z1l231z-gcc-13.3.0-lib/lib/libstdc++.so.6 (0x00007fea68000000)
	libm.so.6 => /nix/store/3dyw8dzj9ab4m8hv5dpyx7zii8d0w6fi-glibc-2.39-52/lib/libm.so.6 (0x00007fea682c9000)
	libgcc_s.so.1 => /nix/store/22nxhmsfcv2q2rpkmfvzwg2w5z1l231z-gcc-13.3.0-lib/lib/libgcc_s.so.1 (0x00007fea682a4000)
	libc.so.6 => /nix/store/3dyw8dzj9ab4m8hv5dpyx7zii8d0w6fi-glibc-2.39-52/lib/libc.so.6 (0x00007fea67e09000)
	/nix/store/3dyw8dzj9ab4m8hv5dpyx7zii8d0w6fi-glibc-2.39-52/lib64/ld-linux-x86-64.so.2 (0x00007fea683be000)
```

**Reply by SuperBo** [reply]
for the use case of Nixos, because library path is not standard, for luagit2 to work, you have to compile lua binding when installing plugin. Then it defeat the orignal purpose of a compiled binary distributed via luarocks (that you suggested in previous comment).

**Reply by Freed-Wu** [reply]
Yes, Nixos is one of unusual OSes :smile:
